### PR TITLE
chore(lint): re-enable `@typescript-eslint/ban-types`

### PR DESCRIPTION
**Pull request by KATT**
Context / Background: https://twitter.com/JoshuaKGoldberg/status/1606070961182228481

## 🎯 Changes

- Re-add `@typescript-eslint/ban-types`
- Add `eslint-ignore` on any interop files
- Fix the rest

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "dev-www": "turbo --filter www dev",
     "test": "turbo --filter \"@trpc/tests\" test",
     "test-watch": "cd packages/tests && pnpm test-watch",
-    "lint": "eslint --ext \".js,.ts,.tsx\" --ignore-path .gitignore .",
+    "lint": "eslint --ext \".js,.ts,.tsx\" --ignore-path .gitignore packages",
```
Comment (KATT) [inline package.json:15]:
```suggestion
    "lint": "eslint --ext \".js,.ts,.tsx\" --ignore-path .gitignore .",
```

**Review by sachinraja — APPROVED** [review]
looks good! had some concerns about if this would affect what TypeScript shows in the tooltip when you hover over a variable (the string representation of a type) but that seems to still be fine

File: packages/tests/server/interop/index.test.tsx [hunk 2]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/ban-types */
```
Comment (JoshuaKGoldberg) [inline packages/tests/server/interop/index.test.tsx:1]: Any particular reason for this one? I tried switching the `{}`s to `object`s in it and didn't see any build failures.

File: packages/tests/server/interop/react/__testHelpers.tsx [hunk 3]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/ban-types */
```
Comment (JoshuaKGoldberg) [inline packages/tests/server/interop/react/__testHelpers.tsx:1]: Any particular reason for this one? I tried switching the `{}` to `object` in it and didn't see any build failures.

File: packages/tests/server/interop/regression/issue-949-inferProcedureInput.test.ts [hunk 4]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/ban-types */
```
Comment (JoshuaKGoldberg) [inline packages/tests/server/interop/regression/issue-949-inferProcedureInput.test.ts:1]: Any particular reason for this one? I tried switching the `{}`s to `object`s in it and didn't see any build failures.

**Review by JoshuaKGoldberg — COMMENTED** [review]
Lovely, thank you for taking action on this! It's looking great. ☺️

Just left a few comments inline where files that you might be able to undo the `eslint-disable`s in.

**Pull request merged**